Content.Server/GameObjects/Components/Medical/CloningMachineComponent.cs

```diff
@@ -0,0 +1,156 @@
+using Content.Server.GameObjects.Components.Observer;
+using Content.Server.GameObjects.Components.Power.ApcNetComponents;
+using Content.Server.GameObjects.EntitySystems;
+using Content.Server.Interfaces;
+using Content.Server.Mobs;
+using Content.Shared.GameObjects.Components.Damage;
+using Content.Shared.GameObjects.Components.Medical;
+using Content.Shared.Interfaces.GameObjects.Components;
+using Content.Shared.Preferences;
+using Robust.Server.GameObjects;
+using Robust.Server.GameObjects.Components.Container;
+using Robust.Server.GameObjects.Components.UserInterface;
+using Robust.Server.Interfaces.GameObjects;
+using Robust.Server.Interfaces.Player;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Interfaces.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Network;
+
+namespace Content.Server.GameObjects.Components.Medical
+{
+    [RegisterComponent]
+    [ComponentReference(typeof(IActivate))]
+    public class CloningMachineComponent : SharedCloningMachineComponent, IActivate
+    {
+        private AppearanceComponent _appearance;
+        private BoundUserInterface _userInterface;
+        private ContainerSlot _bodyContainer;
+        private CloningMachineStatus _status;
+        [Dependency] private readonly IServerPreferencesManager _prefsManager;
+
+        private PowerReceiverComponent _powerReceiver;
+        private Mind _capturedMind;
+        private bool Powered => _powerReceiver.Powered;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            _appearance = Owner.GetComponent<AppearanceComponent>();
+            _userInterface = Owner.GetComponent<ServerUserInterfaceComponent>()
+                .GetBoundUserInterface(CloningMachineUIKey.Key);
+            _userInterface.OnReceiveMessage += OnUiReceiveMessage;
+            _bodyContainer = ContainerManagerComponent.Ensure<ContainerSlot>($"{Name}-bodyContainer", Owner);
+            _powerReceiver = Owner.GetComponent<PowerReceiverComponent>();
+
+            //TODO: write this so that it checks for a change in power events and acts accordingly.
+            var newState = GetUserInterfaceState();
+            _userInterface.SetState(newState);
+
+            UpdateUserInterface();
+        }
+
+        public void Update(float frametime)
+        {
+            UpdateUserInterface();
+            if (_bodyContainer != null && _capturedMind != null)
+            {
+                if (_capturedMind.ReturnToCloning)
+                {
+                    //When the user has confirmed return intent and we are cloning them
+                    _capturedMind.VisitingEntity.Delete();
+                    _capturedMind.TransferTo(_bodyContainer.ContainedEntity);
+                    _bodyContainer.Remove(_bodyContainer.ContainedEntity);
+                    _capturedMind.ReturnToCloning = false;
+                    _capturedMind = null;
+                }
+            }
```

Comment:
Not happy with this being an update loop.

Can't you have the ghost send the machine a message? via the event bus maybe?

Reply:
I'll look into it. I also wasn't really satisfied with this.

Reply:
I'll look into the event bus. I wasn't satisfied with a quasi flag either.